# NumPy

## 1. NumPy Section Introduction

### Vectors and Matrices
- In Linear Algebra, convention to treat vectors as 2-D (column vector (dx1), or raw vector (1xd))
- But not in NumPy! Vectors will be 1-D (most of the time)

### Dot Product / Inner Product
$ a . b = a^T b = \sum_{d=1}^{D} a_d b_d $

### Matrix Multiplication (C = AB)
- Generalized dot product.
- $ c_{ij} = a_{i1}b_{1j} + a_{i2}b_{2j} + ... + a_{in}b_{nj} = \sum_{k=1}^{n} a_{ik}b_{kj} $

### Element-wise Product
- Not so common in Linear Algebra, but very common in ML.

### And many more
- Linear Systems: Ax = b
- Inverse: $ A^{-1} $
- Determinant: $ |A| $
- Choosing random numbers (e.g. Uniform, Gaussian)

### Applications
1. Linear Regression
1. LogisticRegression
1. Deep Neural Networks
1. K-Means Clustering
1. Density Estimation
1. Principal Components Analysis
1. Matrix Factorization (Recommender Systems)
1. Support Vector Machines (SVM)
1. Markov Models and HMM
1. Control Systems
1. Game Theory
1. Operations Research
1. Portofolio Optimization

## 2. NumPy Arrays vs. Python Lists
- Python List looks like an array, but it works like a generic data structure.
- NumPy Array exists specifically to do Math

### Questions

1. Import NumPy Library
1. Creating a list containing 1,2,3
1. Creating NumPy array containing the same items
1. Iterate over a list elements and print them
1. Iterate over a NumPy array elements and print them 
1. Add a new item to a list to the end
1. Size of NumPy array is fixed (immutable), you cannot add an element to NumPy array. But, new array will be created.
1. Add 2 lists together (list concatenation)
1. Add array of one number to an array (broadcasting)
1. Add 2 arrays of the same size (array addition)
1. Multiply a list with a scalar value (list repetition). Multiply operator does repetition.
1. Array Scaling (multiply an array witha scalar value). Multiply operator does scaling.
1. Add a value 3 to each element in a list using for-loop.
1. Add a value 3 to each element in a list using list comprehension.
1. Square each element in a list using list comprehension 
1. Square, SQRT, Log, Exponential, Hyperbolic Tangent each element in NumPy array 

In [1]:
import numpy as np

In [2]:
x = [1,2,3]
print(x)

[1, 2, 3]


In [3]:
y = np.array([1,2,3])
print(y)

[1 2 3]


In [4]:
for element in x:
    print(element)

1
2
3


In [5]:
for element in y:
    print(element)

1
2
3


In [6]:
x.append(4)
print(x)

[1, 2, 3, 4]


In [7]:
x + [1,2,3]

[1, 2, 3, 4, 1, 2, 3]

In [8]:
y + np.array([4])

array([5, 6, 7])

In [9]:
y + np.array([1,2,3])

array([2, 4, 6])

In [10]:
x * 2

[1, 2, 3, 4, 1, 2, 3, 4]

In [11]:
y * 2

array([2, 4, 6])

In [12]:
x2 = []
for element in x:
    x2.append(element + 3)
print(x2)    

[4, 5, 6, 7]


In [13]:
x2 = [element+3 for element in x]
print(x2)

[4, 5, 6, 7]


In [14]:
x3 = [element**2 for element in x]
print(x3)

[1, 4, 9, 16]


In [15]:
# Square, SQRT, Log, Exponential, Hyperbolic Tangent for NumPy Array
print(y**2)
print(np.sqrt(y))
print(np.log(y))
print(np.exp(y))
print(np.tanh(y))

[1 4 9]
[1.         1.41421356 1.73205081]
[0.         0.69314718 1.09861229]
[ 2.71828183  7.3890561  20.08553692]
[0.76159416 0.96402758 0.99505475]


## 3. Dot Product

- It is also called inner product.
- $ a.b = a^T b = \sum^{D}_{d=1}a_d b_d $
- $ a.b = a^T b = \lVert a \rVert \lVert b \rVert \cos \theta_{ab} $
- $ \cos \theta_{ab} = \frac{a^T b}{\lVert a \rVert \lVert b \rVert} $
- $ \lVert a \rVert = \sqrt{\sum_{d=1}^{D} a^{2}_{d}} $

### Questions
1. Declare 2 of 1-D NumPy arrays, each one of them has 3 elements.
1. Determine the dot product using explicit for-loop by 3 methods.
1. Determine the dot product using NumPy (function or instance method).
1. Determine the theta between the 2 vectors in degree form.

In [16]:
import numpy as np

a = np.array([1,2,3])
b = np.array([4,5,6])

dot_product = 0

for element_a, element_b in zip(a,b):
    dot_product += element_a * element_b

print(dot_product)    

32


In [17]:
dot_product = 0

for i in range(len(a)):
    dot_product += a[i] * b[i]

print(dot_product)    

32


In [18]:
dot_product = np.sum(a*b)
print(dot_product)

dot_product = (a*b).sum()
print(dot_product)

32
32


In [19]:
print(a.dot(b))

32


In [20]:
dot_product = np.dot(a,b)
print(dot_product)

dot_product = a.dot(b)
print(dot_product)

dot_product = a @ b
print(dot_product)

32
32
32


In [21]:
# a_magnitude = np.linalg.norm(a)
a_magnitude = a.dot(a)
b_magnitude = b.dot(b)
cos_theta = a.dot(b) / (a_magnitude * b_magnitude)
theta = np.arccos(cos_theta) * 180 / np.pi
print(theta)

88.29894775634656


## 4. Speed Test
### Questions
1. Compare the execution time between Dot Product calculations using:
    1. Explicit for loop on Python List.
    1. Build-in functions in NumPy Array.
1. Compare the execution time between squaring elements of a list:
    1. Explicit for loop on Python List. 
    1. Python list comprehension.

In [22]:
import numpy as np
from datetime import datetime

def dot_product_using_python_list(a,b):
    
    dot_product = 0
    for element_a, element_b in zip(a,b):
        dot_product += (element_a * element_b)
    return dot_product    

a = np.random.randn(100)
b = np.random.randn(100)

T = 100000

before = datetime.now()
for i in range(T):
    dot_product_using_python_list(a,b)
after = datetime.now()
slow_execution_time = after - before

before = datetime.now()
for i in range(T):
    np.dot(a,b)
after = datetime.now()
fast_execution_time = after - before

print(slow_execution_time / fast_execution_time)

11.451574812637059


## 5. Matrices
1. NumPy Array is better to be used than NumPy Matrix.
1. Because NumPy Array can be any dimension, but NumPy Matrix has to be 2-D array.
1. Also, NumPy Matrix has to be converted to NumPy Array first before doing any processing.

### Questions
1. Declare a 2x2 matrix using Python List.
1. Print the first row using one statement.
1. Print one element.

### Questions 
1. Declare a 3x3 array using NumPy Array.
1. Print a certain element from that array using [ ][ ] and [ , ]
1. Print a certain column using a colon notation.
1. Apply matrix transpose.
1. Apply element-wise exponential on the array.
1. Apply element-wise exponential on the NumPy 2d list. It will return a NumPy array.
1. Apply element-wise multiplication between 2 2-D NumPy arrays.
1. Apply Matrix multiplication between 2 2-D NumPy arrays (dot(), @).
1. Apply determinant of an array (np.linalg.det()).
1. Apply determinant of an array (np.linalg.inv()). Check the answer is correct.
1. Apply trace of an array.
1. Get the diagonal elements of an array (matrix to vector).
1. Construct a diagonal matrix of a vector (vector to matrix).
1. Get the eigen values and vectors (np.linalg.eig()) of a matrix, and check the validity of eigen value decomposition equation using np.allclose() function. 
- To Get the eigen values and vectors of a symmetric matrix, use (np.linalg.eigh()). And, to check the validity of eigen value decomposition equation, use (np.allclose()) function.

In [23]:
import numpy as np

x = np.array([[1,2,3], [4,5,6], [7,8,9]])
x

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [24]:
print(x[2][1])

8


In [25]:
x[2,1]

8

In [26]:
x[:,1]

array([2, 5, 8])

In [27]:
x.T

array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

In [28]:
np.exp(x)

array([[2.71828183e+00, 7.38905610e+00, 2.00855369e+01],
       [5.45981500e+01, 1.48413159e+02, 4.03428793e+02],
       [1.09663316e+03, 2.98095799e+03, 8.10308393e+03]])

In [29]:
a = [[1,2,3], [4,5,6], [7,8,9]]
np.exp(a)

array([[2.71828183e+00, 7.38905610e+00, 2.00855369e+01],
       [5.45981500e+01, 1.48413159e+02, 4.03428793e+02],
       [1.09663316e+03, 2.98095799e+03, 8.10308393e+03]])

In [30]:
y = np.array([[1,2,3], [4,5,6], [7,8,9]])
x * y

array([[ 1,  4,  9],
       [16, 25, 36],
       [49, 64, 81]])

In [31]:
x.dot(y)

array([[ 30,  36,  42],
       [ 66,  81,  96],
       [102, 126, 150]])

In [32]:
z = np.array([[1,2], [3,4]])
np.linalg.det(z)

-2.0000000000000004

In [33]:
np.linalg.inv(z)

array([[-2. ,  1. ],
       [ 1.5, -0.5]])

In [34]:
z.dot(np.linalg.inv(z)) 

array([[1.00000000e+00, 1.11022302e-16],
       [0.00000000e+00, 1.00000000e+00]])

In [35]:
z.trace()

5

In [36]:
np.diag(x)

array([1, 5, 9])

In [37]:
a = np.array([1,2,3])
np.diag(a)

array([[1, 0, 0],
       [0, 2, 0],
       [0, 0, 3]])

In [38]:
eig_values, eig_vectors = np.linalg.eig(x)
print(eig_values)
print(eig_vectors)

[ 1.61168440e+01 -1.11684397e+00 -3.38433605e-16]
[[-0.23197069 -0.78583024  0.40824829]
 [-0.52532209 -0.08675134 -0.81649658]
 [-0.8186735   0.61232756  0.40824829]]


In [39]:
np.allclose(x.dot(eig_vectors), eig_vectors @ np.diag(eig_values))

True

## 6. Solving Linear Systems
### Linear Systems: Example Problem

The admission fee at a small fair is \\$1.50 for each child and \\$4.00 for each adult. On a certain day, 2200 people enter the fair, and \\$5050 is collected. How many children and how many adults attended?

2 equations, and 2 unknowns.

- $ x_1 + x_2 = 2200 $
- $ 1.5 x_1 + 4 x_2 = 5050 $

### Do not do that literally!
- The inverse is slower and less accurate.
- There are better algorithms to solve linear systems.

```
x = np.linalg.solve(A, b) # yes
x = np.linalg.inv(A).dot(b) # no
```

In [40]:
import numpy as np

A = np.array([[1, 1], [1.5, 4]])
b = np.array([2200, 5050])

np.linalg.solve(A, b)

array([1500.,  700.])

## 7. Generating Data

### Questions
1. Create an array of zeros of size (2x3).
1. Create an array of ones of size (2x3).
1. Create an array of tens of size (2x3).
1. Create an identity matrix of size (3x3).
1. Generate a random number using random module.
1. Generate a random matrix of size (2x3) from the Uniform Distribution [0,1].
1. Generate a random matrix of size (2x3) from the Normal Distribution (mean=0, variance=1).
1. Generate a random vector of size (10,000) from the Normal Distribution (mean=0, variance=1).
1. Calculate the mean of this vector.
1. Calculate the variance of this vector.
1. Calculate the Standard Deviation of this vector. Chect the correctness of the output.
1. Generate a random matrix of size (10,000x3) (observations, measurements) from the Normal Distribution (mean=0, variance=1).
1. Calculate the mean of each column.
1. Calculate the mean of each row. Print its shape.
1. Calculate the Covarience matrix of the matrix transpose. Print its shape.
1. Calculate the Covarience matrix of the matrix using rowvar=False. Print its shape.
1. Generate a random int matrix of size (3x3) within [0, 10[
1. Use choice function to randomly choose (3x3) numbers from [0, 10[

In [41]:
import numpy as np

np.zeros((2,3))

array([[0., 0., 0.],
       [0., 0., 0.]])

In [42]:
np.ones((2,3))

array([[1., 1., 1.],
       [1., 1., 1.]])

In [43]:
10 * np.ones((2,3))

array([[10., 10., 10.],
       [10., 10., 10.]])

In [44]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [45]:
np.random.random()

0.30087709586783795

In [46]:
np.random.random((2,3))

array([[0.55615787, 0.11562383, 0.73755509],
       [0.95297875, 0.23153665, 0.1814484 ]])

In [47]:
np.random.randn(2,3)

array([[-0.57484303, -0.6040536 , -0.60025974],
       [-0.9226913 ,  1.7439405 ,  2.16489853]])

In [48]:
x = np.random.randn(10000)

In [49]:
np.mean(x)

-0.011245871539846991

In [50]:
np.var(x)

1.0013840909392984

In [51]:
np.std(x)

1.0006918061717596

In [52]:
np.std(x) == np.sqrt(np.var(x))

True

In [53]:
np.allclose(np.std(x), np.sqrt(np.var(x)))

True

In [54]:
x = np.random.randn(10000, 3)

In [55]:
np.mean(x, axis=0)

array([ 0.00765856, -0.00236475,  0.00113244])

In [56]:
np.mean(x, axis=1).shape

(10000,)

In [57]:
np.cov(x.T)

array([[ 0.99077244,  0.00487418, -0.01392852],
       [ 0.00487418,  1.01437346, -0.00199158],
       [-0.01392852, -0.00199158,  1.01550615]])

In [58]:
np.cov(x.T).shape

(3, 3)

In [59]:
np.cov(x, rowvar=False)

array([[ 0.99077244,  0.00487418, -0.01392852],
       [ 0.00487418,  1.01437346, -0.00199158],
       [-0.01392852, -0.00199158,  1.01550615]])

In [60]:
np.cov(x, rowvar=False).shape

(3, 3)

In [61]:
np.random.randint(0, 10, size=(3,3))

array([[2, 9, 9],
       [5, 1, 6],
       [6, 2, 9]])

In [62]:
np.random.choice(np.arange(10), size=(3,3))

array([[1, 6, 6],
       [7, 2, 3],
       [7, 0, 2]])

In [63]:
np.random.choice(10, size=(3,3))

array([[5, 4, 4],
       [0, 5, 4],
       [5, 1, 8]])